pageserver/src/storage_sync.rs

```diff
@@ -129,1330 +114,587 @@
+        let mut upload_queue = self.upload_queue.lock().unwrap();
+        ensure!(
+            upload_queue.latest_metadata.is_some(),
+            "upload queue not initialized"
+        );
 
-async fn download_timeline_data(
-    conf: &'static PageServerConf,
-    (storage, index, sync_queue): (&GenericRemoteStorage, &RemoteIndex, &SyncQueue),
-    current_remote_timeline: Option<&RemoteTimeline>,
-    sync_id: TenantTimelineId,
-    new_download_data: SyncData<LayersDownload>,
-    sync_start: Instant,
-) -> DownloadStatus {
-    static TASK_NAME: &str = "download";
-
-    match download_timeline_layers(
-        conf,
-        storage,
-        sync_queue,
-        current_remote_timeline,
-        sync_id,
-        new_download_data,
-    )
-    .await
-    {
-        DownloadedTimeline::Abort => {
-            register_sync_status(sync_id, sync_start, TASK_NAME, None);
-            if let Err(e) = index.write().await.set_awaits_download(&sync_id, false) {
-                error!("Timeline {sync_id} was expected to be in the remote index after a download attempt, but it's absent: {e:?}");
-            }
-        }
-        DownloadedTimeline::FailedAndRescheduled => {
-            register_sync_status(sync_id, sync_start, TASK_NAME, Some(false));
+        // Update the remote index file, removing the to-be-deleted files from the index,
+        // before deleting the actual files.
+        for path in paths {
+            let relative_path = RelativePath::from_local_path(
+                &self.conf.timeline_path(&self.timeline_id, &self.tenant_id),
+                path,
+            )?;
+            upload_queue.latest_files.remove(&relative_path);
         }
-        DownloadedTimeline::Successful(mut download_data) => {
-            match update_local_metadata(conf, sync_id, current_remote_timeline).await {
-                Ok(()) => match index.write().await.set_awaits_download(&sync_id, false) {
-                    Ok(()) => {
-                        register_sync_status(sync_id, sync_start, TASK_NAME, Some(true));
-                        return DownloadStatus::Downloaded;
-                    }
-                    Err(e) => {
-                        error!("Timeline {sync_id} was expected to be in the remote index after a successful download, but it's absent: {e:?}");
-                    }
-                },
-                Err(e) => {
-                    error!("Failed to update local timeline metadata: {e:?}");
-                    download_data.retries += 1;
-                    sync_queue.push(sync_id, SyncTask::Download(download_data));
-                    register_sync_status(sync_id, sync_start, TASK_NAME, Some(false));
-                }
-            }
+        let disk_consistent_lsn = upload_queue
+            .latest_metadata
+            .as_ref()
+            .unwrap()
+            .disk_consistent_lsn();
+        let index_part = IndexPart::new(
```

Comment:
Could be extracted into a separate method in the queue, that generates current `IndexPart`.

Reply:
It could, but I think it's fine as it is.